### PR TITLE
Allow use of arbitrary account IDs

### PR DESCRIPTION
**Pull request by viren-nadkarni**
### Summary

This PR introduces the ability to use resource namespacing based on Account IDs in LocalStack community edition.

### Related

Docs: https://github.com/localstack/docs/pull/323

See: https://github.com/localstack/localstack/issues/7068

### CODE REVIEW

**Review by thrau — APPROVED** [review]
love that we're getting this into community :partying_face: just a minor comment regarding logging, but feel free to merge.

File: localstack/aws/accounts.py [hunk 1]
```diff
@@ -31,30 +36,37 @@ def set_aws_access_key_id(access_key_id: str):
 
 def get_aws_account_id() -> str:
     """Return the AWS account ID for the current context."""
-    return account_id_resolver()
+    try:
+        return REQUEST_CTX_TLS.account_id
+    except AttributeError:
+        if LS_LOG and LS_LOG == LS_LOG_TRACE_INTERNAL:
+            LOG.debug(
```
[the remark this answers is not in the document]
Comment (viren-nadkarni) [reply localstack/aws/accounts.py]: Let's keep it consistent. Fixed in 8477bd1 :+1:

File: localstack/aws/proxy.py [hunk 2]
```diff
@@ -28,7 +32,11 @@ def get_account_id_from_request(request: Request) -> str:
         extract_access_key_id_from_auth_header(request.headers) or TEST_AWS_ACCESS_KEY_ID
     )
     set_aws_access_key_id(access_key_id)
-    return get_account_id_from_access_key_id(access_key_id)
+
+    account_id = get_account_id_from_access_key_id(access_key_id)
+    set_aws_account_id(account_id)
```
Comment (dominikschubert) [inline localstack/aws/proxy.py:37]: IMO the `set_*` calls should be extracted. Would make this easier to test in isolation and more predictable in terms of behavior.

File: localstack/aws/handlers/auth.py [hunk 3]
```diff
@@ -33,18 +37,21 @@ class AccountIdEnricher(Handler):
...
-        access_key_id = extract_access_key_id_from_auth_header(context.request.headers)
-
-        if not access_key_id:
-            access_key_id = TEST_AWS_ACCESS_KEY_ID
-
-        # Save the request access key ID in the current thread local storage
+        # Obtain the access key ID and save it in the thread context
+        access_key_id = (
+            extract_access_key_id_from_auth_header(context.request.headers)
+            or TEST_AWS_ACCESS_KEY_ID
```
Comment (dominikschubert) [inline localstack/aws/handlers/auth.py:43]: `or TEST_AWS_ACCESS_KEY_ID`

When is the fallback here necessary? Are there any reasonable cases that need this?

File: localstack/aws/accounts.py [hunk 4]
```diff
@@ -15,9 +15,9 @@
 #
 
 
-def get_aws_access_key_id() -> Optional[str]:
+def get_aws_access_key_id() -> str:
     """Return the AWS access key ID for current context."""
-    return getattr(REQUEST_CTX_TLS, "access_key_id", None)
+    return getattr(REQUEST_CTX_TLS, "access_key_id", TEST_AWS_ACCESS_KEY_ID)
```
Comment (dominikschubert) [inline localstack/aws/accounts.py:20]: Curious about our strategy here. Should the TLS always have an entry for this? Is this something we *want* to be the case?

File: localstack/aws/accounts.py [hunk 5]
```diff
@@ -31,30 +31,36 @@ def set_aws_access_key_id(access_key_id: str):
...
+def set_aws_account_id(account_id: str) -> None:
+    REQUEST_CTX_TLS.account_id = account_id
 
 
 def get_account_id_from_access_key_id(access_key_id: str) -> str:
     """Return the Account ID associated the Access Key ID."""
-    # This utility ignores IAM mappings.
-    # For now, we assume the client sends Account ID in Access Key ID field.
 
+    # If AWS_ACCES_KEY_ID has a 12-digit integer value, use it as the account ID
```
Comment (dominikschubert) [inline localstack/aws/accounts.py:51]:
```suggestion
    # If AWS_ACCESS_KEY_ID has a 12-digit integer value, use it as the account ID
```

File: localstack/aws/accounts.py [hunk 6]
```diff
@@ -31,30 +31,36 @@ def set_aws_access_key_id(access_key_id: str):
...
+
+    elif len(access_key_id) >= 20:
+        # If AWS_ACCESS_KEY_ID has production AWS credentials, ignore them
+        if access_key_id.startswith("ASIA") or access_key_id.startswith("AKIA"):
+            LOG.warning(
+                "Ignoring production AWS credentials provided to LocalStack. Falling back to default account ID."
+            )
+
+        elif access_key_id.startswith("LSIA") or access_key_id.startswith("LKIA"):
+            # TODO Add IAM mapping logic here
```
Comment (dominikschubert) [inline localstack/aws/accounts.py:63]: cc @dfangl

**Review by dominikschubert — APPROVED** [review]
Minor nits and questions, otherwise LGTM :+1:

File: localstack/aws/handlers/auth.py (hunk 3 above)
Comment (viren-nadkarni) [reply localstack/aws/handlers/auth.py:43]: IIRC there are certain S3 calls where `extract_access_key_id_from_auth_header()` does not apply. I don't remember off the top of my head but can look up if needed

File: localstack/aws/accounts.py (hunk 4 above)
Comment (viren-nadkarni) [reply localstack/aws/accounts.py:20]: Yes, ideally we would want to have the TLS to have the default access key ID. This lazy approach is simply cheaper. Note that this is chiefly used by the `aws_stack` and we'll have to rethink this as we progress in `aws_stack` reworking.

File: localstack/aws/proxy.py (hunk 2 above)
Comment (viren-nadkarni) [reply localstack/aws/proxy.py:37]: Not sure I understand -- could you clarify what you mean by extracting the `set_*` calls?

Comment (viren-nadkarni) [reply localstack/aws/proxy.py:37]: Fair point. The name `get_account_id_from_request` gives the wrong impression. Will fix it separately, thanks :+1:

File: localstack/aws/accounts.py [hunk 7]
```diff
@@ -31,30 +31,36 @@ def set_aws_access_key_id(access_key_id: str):
...
     if re.match(r"\d{12}", access_key_id):
         return access_key_id
-    else:
-        return get_default_account_id()
+
+    elif len(access_key_id) >= 20:
+        # If AWS_ACCESS_KEY_ID has production AWS credentials, ignore them
+        if access_key_id.startswith("ASIA") or access_key_id.startswith("AKIA"):
+            LOG.warning(
+                "Ignoring production AWS credentials provided to LocalStack. Falling back to default account ID."
```
Comment (joebowbeer) [inline localstack/aws/accounts.py:59]: @viren-nadkarni Is there any way to silence or change the level of this new warning?

Comment (viren-nadkarni) [reply localstack/aws/accounts.py:59]: @joebowbeer You could set this config var: `PARITY_AWS_ACCESS_KEY_ID=1`

Comment (joebowbeer) [reply localstack/aws/accounts.py:59]: Thanks for the suggestion and doc. I wish there were a way to ignore genuine creds `PARITY_AWS_ACCESS_KEY_ID=0` _and_ silence the new warning.

**Pull request merged**